File: my-folder/0110-balanced-binary-tree/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isBalanced(root: TreeNode | null): boolean {
    if (root === null) return true;

    const ld = maxDepth(root.left, 1);
    const rd = maxDepth(root.right, 1);

    return Math.abs(ld - rd) > 1? false: isBalanced(root.left) && isBalanced(root.right);
};

function maxDepth(root: TreeNode | null, depth: number): number {
    if (root === null) return depth;

    const ld = maxDepth(root.left, depth + 1);
    const rd = maxDepth(root.right, depth + 1);

    return (ld > rd)? ld: rd;
}

File: my-folder/0110-balanced-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def max_height(root: Optional[TreeNode], height: int) -> int:
            if not root:
                return height
            
            return max(max_height(root.left, height + 1), max_height(root.right, height + 1))
        
        if not root:
            return True

        l = max_height(root.left, 1)
        r = max_height(root.right, 1)
        diff = abs(l - r)
        
        return diff < 2 and self.isBalanced(root.left) and self.isBalanced(root.right)
